woocommerce/woocommerce-android
# Format currencies based on site settings and locale

Resolves #148, improving the formatting and localization of currencies in the app.

**Relies on https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1094, which should be reviewed and merged before this PR.**

Currencies in the dashboard and order views are now formatted based on two things:

#### 1. The site's currency settings

Specifically, the currency symbol position (left, right, left with space, right with space), the thousands divider, the decimal point divider, and the number of decimal points:

![Screen Shot 2018-02-08 at 5.18.06 PM.png](https://images.zenhubusercontent.com/5a0ae8908a75884b908759ea/c1b5c9d2-0471-4377-bd6e-356530efef90)

These are only currently applied for values in the order views. The dashboard uses a rounded format (and also uses `1.2k` for thousands, for example), which skips the currency display settings (though the currency code is still resolved to a symbol, as described in the step below).

#### 2. The currency code

This is either the currency code for the order, or the currency code of the stats in the dashboard. If no currency code exists in either case, we'll fall back to the global currency code from the site's settings (shown above in `wp-admin`) - and if that doesn't exist either, fall back to the default currency for the device's locale.

Whatever the case, the currency symbol for the currency code will be localized to the device's locale. For example, in the US, USD are shown as `$`. In Canada, USD are shown as `US$`.

### Notes

I tried something a little different for making the formatter available to the various nested views we have. I went for creating a transformation function that can be used, and passing it to the nested views from the top-level fragments. This hides some of the internals from the nested views, and simplifies what the formatting calls look like at the time they're made.

Update release notes:


#### To test

1. Play around with your device's locale, and your WooCommerce site's currency settings in wp-admin under WooCommerce -> Settings (note that restarting the app is the easiest way to trigger an updated of the site settings - see [here](https://github.com/woocommerce/woocommerce-android/blob/38a2c2b598b2fcfdf26fa666d2b9bc30dcbb035c/WooCommerce/src/main/kotlin/com/woocommerce/android/WooCommerce.kt#L139)
2. Check the case where the site settings aren't stored by disabling the update call [here](https://github.com/woocommerce/woocommerce-android/blob/38a2c2b598b2fcfdf26fa666d2b9bc30dcbb035c/WooCommerce/src/main/kotlin/com/woocommerce/android/WooCommerce.kt#L139), and then clearing the `WCSiteModel` table via Stetho
